lib/messageReceiver.ts

```diff
@@ -163,29 +164,99 @@ export class MessageReceiver extends LinkEntity {
+        // - This autorenewal needs to happen **NO MORE** than maxAutoRenewDurationInSeconds
+        // - We should be able to clear the renewal timer when the user's message handler
+        // is done (whether it succeeds or fails).
+        log.receiver("[%s] message with id '%s' is locked until %s.",
+          connectionId, bMessage.messageId, bMessage.lockedUntilUtc!.toString());
+        const totalAutoLockRenewDuration = Date.now() + (this.maxAutoRenewDurationInSeconds * 1000);
+        log.receiver("[%s] Total autolockrenew duration for message with id '%s' is: ",
+          connectionId, bMessage.messageId, new Date(totalAutoLockRenewDuration).toString());
+        const autoRenewLockTask = (): void => {
+          if (Date.now() < totalAutoLockRenewDuration && continueExecution) {
+            const amount = calculateRenewAfterDuration(bMessage.lockedUntilUtc!);
+            log.receiver("[%s] Sleeping for %d milliseconds while renewing the lock for message " +
+              "with id '%s' is: ", connectionId, amount, bMessage.messageId);
+            timer = setTimeout(async () => {
+              try {
+                log.receiver("[%s] Attempting to renew the lock for message with id '%s'.",
+                  connectionId, bMessage.messageId);
+                await this._context.managementClient!.renewLock(bMessage);
+                log.receiver("[%s] Successfully renewed the lock for message with id '%s'.",
+                  connectionId, bMessage.messageId);
+              } catch (err) {
+                log.error("[%s] An error occured while auto renewing the message lock '%s' for " +
+                  "message with id '%s': %O.", connectionId, bMessage.lockToken,
+                  bMessage.messageId, err);
+              }
+              log.receiver("[%s] Calling the autorenewlock task again for message with id '%s'.",
+                connectionId, bMessage.messageId);
+              autoRenewLockTask();
+            }, amount);
+          }
+        };
+        // start
+        autoRenewLockTask();
+      }
       try {
         await this._onMessage(bMessage);
-        if (this.autoComplete) {
-          log[this.receiverType]("[%s] Auto completing the message with id '%s' on the receiver '%s'.",
-            connectionId, bMessage.messageId, this.name);
-          await bMessage.complete();
-        }
+        continueExecution = false;
+        clearTimeout(timer);
+        log.receiver("[%s] Stopping lock renewal for message with id '%s'.", connectionId,
+          bMessage.messageId);
       } catch (err) {
-        log.error("[%s] Abandoning the message with id '%s' on the receiver '%s' since an error " +
-          "occured: %O.", connectionId, bMessage.messageId, this.name);
+        // Do not want renewLock to happen unnecessarily, while abandoning the message. Hence,
+        // doing this here. Otherwise, this should be done in finally.
+        continueExecution = false;
```

Comment:
This is probably DRY overkill, but you could potentially add an inline function before the `try`/`catch` that you call in both places which will clear the timer and set `continueExecution` to false. With only two lines may not be worth it now, but I'd consider it if you start having more logic shared between the two